### DISCUSSION TITLE
(Need help) Can we restart botorch from previous optimization?

### DISCUSSION THREAD

**Discussion by mwn0216** [meta-pytorch/botorch#2466] [Q&A]
Hi, guys, i am new to the botorch. I am using it to optimize some simulation results. Sometimes, the simulation may fails and cause the whole optimization process stoped. Then, I have to re-run all of the optimization. Is there a way to re-start BO with the previous data? For example, if we have run it from 0 iteration to 100 iteration, then some error happens, can we export some data and start the next optimization from the 100 iteration? If this can be done, it will be really appreciated if you can tell me how to do that or where I can find the instruction. Thank you for your generous help.

**Comment by Balandat** [comment]
Sure, this is definitely possible. How exactly are you using BoTorch to do this optimization?

Note that BoTorch is primarily designed for Bayesian Optimization research & advanced use cases; if you simply want to optimize a black box function I'd recommend using Ax (https://ax.dev/), which has out-of-the-box support for storing experiments snapshots.

**Reply by mwn0216** [reply]
Thank you for your response! I am currently optimizing a simulation to get a best performance material. I am using the Ax platfrom to do this simple test case, but may use Botorch in the future. Back to this problem, can you please tell me where I can get the instruction to get out-of-the-box data and restart the optimization from the previous stop point?

The following part is my code, which is a very simple one from the tuturial.

Thank you again for the help.

P.S. If I post in the wrong area, I appology and will repost this question in the Ax section. Perviously I thought Ax and Botorch are the same thing, now I get it.  :)

from botorch.models.gpytorch import GPyTorchModel
from gpytorch.distributions import MultivariateNormal
from gpytorch.kernels import RBFKernel, ScaleKernel
from gpytorch.likelihoods import GaussianLikelihood
from gpytorch.means import ConstantMean
from gpytorch.models import ExactGP
from torch import Tensor
from ax.models.torch.botorch_modular.model import BoTorchModel
from ax.models.torch.botorch_modular.surrogate import Surrogate
from ax.modelbridge.generation_strategy import GenerationStep, GenerationStrategy
from ax.modelbridge.registry import Models
import torch
from ax.service.ax_client import AxClient
from ax.service.utils.instantiation import ObjectiveProperties


(in the middle is my DEM simulation, I define a function named "Triaxial" to generate the simulation and return simulation result back to the optimization)


##############             NUmber of evaluation            ##############################
NUM_EVALS = 200
initialtrials=5
##############             Customize GP model              ##############################
class SimpleCustomGP(ExactGP, GPyTorchModel):

    _num_outputs = 1  # to inform GPyTorchModel API

    def __init__(self, train_X, train_Y, train_Yvar: Optional[Tensor] = None):
        # NOTE: This ignores train_Yvar and uses inferred noise instead.
        # squeeze output dim before passing train_Y to ExactGP
        super().__init__(train_X, train_Y.squeeze(-1), GaussianLikelihood())
        self.mean_module = ConstantMean()
        self.covar_module = ScaleKernel(
            base_kernel=RBFKernel(ard_num_dims=train_X.shape[-1]),
        )
        self.to(train_X)  # make sure we're on the right device/dtype

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return MultivariateNormal(mean_x, covar_x)
##############             CUstomize Botorch Model           #############################
ax_model = BoTorchModel(
    surrogate=Surrogate(
        # The model class to use
        botorch_model_class=SimpleCustomGP,
        # Optional, MLL class with which to optimize model parameters
        # mll_class=ExactMarginalLogLikelihood,
        # Optional, dictionary of keyword arguments to model constructor
        # model_options={}
    ),
    # Optional, acquisition function class to use - see custom acquisition tutorial
    # botorch_acqf_class=qExpectedImprovement,
)
###############     Generation strategy for initial and optimization      ###############
gs = GenerationStrategy(
    steps=[
        # Quasi-random initialization step
        GenerationStep(
            model=Models.SOBOL,
            num_trials=initialtrials,  # How many trials should be produced from this generation step
        ),
        # Bayesian optimization step using the custom acquisition function
        GenerationStep(
            model=Models.BOTORCH_MODULAR,
            num_trials=-1,  # No limitation on how many trials should be produced from this step
            # For `BOTORCH_MODULAR`, we pass in kwargs to specify what surrogate or acquisition function to use.
            model_kwargs={
                "surrogate": Surrogate(SimpleCustomGP),
            },
        ),
    ]
)
##
# Initialize the client - AxClient offers a convenient API to control the experiment
ax_client = AxClient(generation_strategy=gs)

########################    generate  particle parameter list         ####################
parameters = []
param_int = {"name": "x1", "type": "range", "bounds": [1, 5], "value_type": "int"}
parameters.append(param_int)
param_int = {"name": "x2", "type": "range", "bounds": [1, 5], "value_type": "int"}
parameters.append(param_int)
param = {"name": "x3", "type": "range", "bounds": [-p, p]}
parameters.append(param)
param = {"name": "x4", "type": "range", "bounds": [-p, p]}
parameters.append(param)
param = {"name": "x5", "type": "range", "bounds": [-p, p]}
parameters.append(param)
########################    Main body of the parameter modify module ####################    
ax_client.create_experiment(
    name="new_test_experiment",
    #define the number and boundary of the parameters !!!!!!!!!!!!
    parameters=parameters,
    objectives={
        "Triaxial_minimize": ObjectiveProperties(minimize=False),  # define whether we want minimize
    },
)
# change evaluate function

def evaluate(parameters):
    Par_list=[]
    for i in  range(1, 6):
    	pin='x'+str(i)
    	Par_list.append(parameters[pin])
    print('\033[95m' + 'The parameter list for this iteration is: ' + str(Par_list) + '\033[0m')
    return {"Triaxial_minimize": (Triaxial(Par_list,Ballnumber), float("nan"))}

**Comment by Balandat** [comment]
I'm not sure why you want to "Customize GP model", this seems unnecessary (and most likely worse than using the Ax defaults). You should start here: https://ax.dev/tutorials/gpei_hartmann_service.html. You can read about evaluation failures in the same tutorial: https://ax.dev/tutorials/gpei_hartmann_service.html#Special-Cases

**Reply by mwn0216** [reply]
Thank you! Will do that. Thank you for the Patient explanation.
